### DISCUSSION TITLE
Stencil functionality for rlgl.

### DISCUSSION THREAD

**Discussion by dnxa** [raysan5/raylib#2964] [Features]
I have already implemented this myself:
```
#include "rlgl.h"
#include "glfw3.h"

#define GL_ALPHA_TEST	 0x0BC0
#define GL_GREATER   	 0x0204

// Stencil masks
void beginStencil()
{
	rlDrawRenderBatchActive();
	glEnable(GL_STENCIL_TEST);
}

void beginStencilMask()
{
	glColorMask(GL_FALSE, GL_FALSE, GL_FALSE, GL_FALSE);
	glStencilFunc(GL_ALWAYS, 1, 0xFF);
	glStencilOp(GL_KEEP, GL_KEEP, GL_REPLACE);
	glEnable(GL_ALPHA_TEST);
	glAlphaFunc(GL_GREATER, 0.5f);
}

void endStencilMask()
{
	rlDrawRenderBatchActive();	
	glDisable(GL_ALPHA_TEST);	
	glStencilFunc(GL_EQUAL, 1, 0xFF);
	glStencilOp(GL_KEEP, GL_KEEP, GL_KEEP);
	glColorMask(GL_TRUE, GL_TRUE, GL_TRUE, GL_TRUE);
}

void endStencil()
{
	rlDrawRenderBatchActive();
	glDisable(GL_STENCIL_TEST);
}
```

glfw3.h is from Raylib's 'external\glfw\include\GLFW' directory.

You can then use it like so:
```
beginStencil();

  beginStencilMask();
  
  // Draw mask
  
  endStencilMask();
  
  // Draw what you want to be masked

endStencil()
```

It's very useful, flexible, and simple to use for making your game look nice, and I think it should be a part of rlgl.

**Comment by Veslo5** [comment]
Awesome, did you considered making pull request?

**Reply by raysan5** [reply]
This is an advance feature that not many users could need in raylib, I prefer not to add it to main library, in any case it can be added as an example.

**Reply by Veslo5** [reply]
Ok, understandable :)

**Reply by bagomot** [reply]
I needed it, I’ll have to connect opengl.

**Reply by zeroxer** [reply]
> This is an advance feature that not many users could need in raylib, I prefer not to add it to main library, in any case it can be added as an example.

we need it!
